### DISCUSSION TITLE
About best practice, using it in project the first time, I've written tons of boilerplate, wondering if it can be more concise?

### DISCUSSION THREAD

**Discussion by FateRiddle** [pmndrs/zustand#754] [Q&A] [answered]
For instance my App component needs to get access to a number of states and methods, I wrote them one by one like this in App:

```js
  const els = useStore((state) => state.els);
  const setEls = useStore((state) => state.setEls);
  const setLabelEditing = useStore((state) => state.setLabelEditing);
  const showTool = useStore((state) => state.showTool);
  const showEditor = useStore((state) => state.showEditor);
  const editorId = useStore((state) => state.editorId);
  const editElById = useStore((state) => state.editElById);
```

Is this the best practice? If not what is?

**Comment by dai-shi** [comment, marked as answer]
That looks good to me. Alternatively, you could combine them like in an array and use `shallow`.

If you are looking for another practice, we can define actions outside useStore in advance.
See: https://github.com/pmndrs/zustand/wiki/Practice-with-no-store-actions

**Comment by FateRiddle** [comment]
@dai-shi Thanks. By the way what's the best way, performance wise, to do a selector?
By that I mean, if I've got `items` and `currentId` saved as zustand states, how does a component get access to `currentItem` which is items[currentId]. Do I fetch both `items` and `currentId` from useStore, then do the math in the given component?

I'd imaging if I do that, any change to `items` would result in a rerender of the component, yet the component is supposed to change only when "current item" is changed.

**Reply by dai-shi** [reply]
performance-wise, defining a stable selector outside a component is the best.

```js
const selectCurrentItem = (state) => state.items[state.currentId]

const Component = () => {
  const currentItem = useStore(selectCurrentItem)
  // ...
}
```

**Comment by TheRefineryAU** [comment]
In this [article](https://dev.to/avt/understanding-zustand-a-beginners-guide-with-typescript-4jjo) the author uses this pattern:

const { count, increment, decrement } = useCounterStore();

I haven't seen this anywhere else but it works for me and saves so mcuh boilerplate! Is there any issues/downsides with using this approach over OP's code?

**Reply by dai-shi** [reply]
This is valid, but if there are some other properties, it can cause extra rerenders.

You might be interested in [valtio](https://github.com/pmndrs/valtio) which optimizes this pattern.

**Reply by TheRefineryAU** [reply]
Is there an easy way to tell if i'm causing extra rerenders?

**Reply by dai-shi** [reply]
Use React DevTools.
